File: apps/yawp/assets/app/__tests__/bip39.test.ts
/**
 * Shared BIP-39 + HKDF-SHA256 test vectors. The same fixture is
 * consumed by the Elixir test suite (test/yawp/bip39_test.exs).
 * Byte-for-byte agreement is the invariant.
 */

import {
  entropyToMnemonic,
  mnemonicToSeed,
  validateMnemonic,
} from '../identity/bip39';
import {hkdfSha256} from '../identity/hkdf';
import fixture from '../../../priv/test_vectors/bip39.json';

interface OfficialVector {
  entropy_hex: string;
  mnemonic: string;
  passphrase: string;
  seed_hex: string;
}

interface YawpVector extends OfficialVector {
  master_derived_hex: string;
  bundle_derived_hex: string;
}

function hexToBytes(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return out;
}

function bytesToHex(b: Uint8Array): string {
  return Array.from(b)
    .map(x => x.toString(16).padStart(2, '0'))
    .join('');
}

describe('BIP-39 official English vectors', () => {
  for (const v of fixture.official_vectors as OfficialVector[]) {
    test(`entropy ${v.entropy_hex.slice(0, 16)}…`, () => {
      const entropy = hexToBytes(v.entropy_hex);
      const words = v.mnemonic.split(' ');
      expect(entropyToMnemonic(entropy)).toEqual(words);
      expect(validateMnemonic(words)).toEqual({ok: true});
      const seed = mnemonicToSeed(words, v.passphrase);
      expect(bytesToHex(seed)).toBe(v.seed_hex);
    });
  }
});

describe('Yawp-specific HKDF derivation vectors', () => {
  for (const v of fixture.yawp_vectors as YawpVector[]) {
    test(`entropy ${v.entropy_hex.slice(0, 16)}…`, () => {
      const entropy = hexToBytes(v.entropy_hex);
      const words = entropyToMnemonic(entropy);
      expect(words).toEqual(v.mnemonic.split(' '));
      const seed = mnemonicToSeed(words, v.passphrase);
      expect(bytesToHex(seed)).toBe(v.seed_hex);

      const master = hkdfSha256(
        seed,
        new TextEncoder().encode('yawp-master-v1'),
        new TextEncoder().encode('ed25519-seed'),
        32,
      );
      expect(bytesToHex(master)).toBe(v.master_derived_hex);

      const bundle = hkdfSha256(
        seed,
        new TextEncoder().encode('yawp-bundle-v1'),
        new TextEncoder().encode('chacha20-poly1305'),
        32,
      );
      expect(bytesToHex(bundle)).toBe(v.bundle_derived_hex);
    });
  }
});

describe('validateMnemonic', () => {
  test('rejects a bad checksum', () => {
    const bad = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon'.split(
      ' ',
    );
    expect(validateMnemonic(bad)).toEqual({ok: false, reason: 'bad_checksum'});
  });

  test('rejects an invalid word count', () => {
    expect(validateMnemonic(['abandon', 'abandon'])).toEqual({
      ok: false,
      reason: 'invalid_word_count',
    });
  });

  test('rejects a word not in the dictionary', () => {
    const bad = 'notaword abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'.split(
      ' ',
    );
    expect(validateMnemonic(bad)).toEqual({ok: false, reason: 'unknown_word'});
  });

  test('accepts the all-zero entropy mnemonic', () => {
    const words = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'.split(
      ' ',
    );
    expect(validateMnemonic(words)).toEqual({ok: true});
  });

  test('accepts the all-ones entropy mnemonic', () => {
    const words = 'zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo wrong'.split(
      ' ',
    );
    expect(validateMnemonic(words)).toEqual({ok: true});
  });
});

describe('entropyToMnemonic', () => {
  test('rejects non-128-bit entropy', () => {
    expect(() => entropyToMnemonic(new Uint8Array(8))).toThrow();
  });
});

describe('hkdfSha256 RFC 5869 vector', () => {
  test('test case 1', () => {
    const ikm = hexToBytes('0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b');
    const salt = hexToBytes('000102030405060708090a0b0c');
    const info = hexToBytes('f0f1f2f3f4f5f6f7f8f9');
    expect(bytesToHex(hkdfSha256(ikm, salt, info, 42))).toBe(
      '3cb25f25faacd57a90434f64d0362f2a2d2d0a90cf1a5a4c5db02d56ecc4c5bf34007208d5b887185865',
    );
  });
});

File: apps/yawp/test/yawp/test_support/bip39_test.exs
defmodule Yawp.TestSupport.Bip39Test do
          use ExUnit.Case, async: true

  alias Yawp.TestSupport.Bip39
  alias Yawp.TestSupport.Hkdf

  @fixture_path Path.expand(
                  "../../../priv/test_vectors/bip39.json",
                  __DIR__
                )

  @fixture @fixture_path |> File.read!() |> Jason.decode!()
  @official_vectors @fixture["official_vectors"]
  @yawp_vectors @fixture["yawp_vectors"]
  @japanese_vectors @fixture["japanese_vectors"]

  describe "official BIP-39 English vectors" do
    for {v, idx} <- Enum.with_index(@official_vectors) do
      @v v
      @idx idx
      test "##{@idx} entropy → mnemonic → seed" do
        entropy = Base.decode16!(@v["entropy_hex"], case: :lower)
        words = String.split(@v["mnemonic"], " ")

        assert Bip39.entropy_to_mnemonic(entropy) == words
        assert Bip39.validate_mnemonic(words) == :ok

        seed = Bip39.mnemonic_to_seed(words, @v["passphrase"])
        assert Base.encode16(seed, case: :lower) == @v["seed_hex"]
      end
    end
  end

  describe "Yawp-specific HKDF derivation vectors" do
    for {v, idx} <- Enum.with_index(@yawp_vectors) do
      @v v
      @idx idx
      test "##{@idx} entropy → master_seed + bundle_key" do
        entropy = Base.decode16!(@v["entropy_hex"], case: :lower)
        words = Bip39.entropy_to_mnemonic(entropy)
        assert words == String.split(@v["mnemonic"], " ")

        seed = Bip39.mnemonic_to_seed(words, @v["passphrase"])
        assert Base.encode16(seed, case: :lower) == @v["seed_hex"]

        master = Hkdf.derive(seed, "yawp-master-v1", "ed25519-seed", 32)
        assert Base.encode16(master, case: :lower) == @v["master_derived_hex"]

        bundle = Hkdf.derive(seed, "yawp-bundle-v1", "chacha20-poly1305", 32)
        assert Base.encode16(bundle, case: :lower) == @v["bundle_derived_hex"]
      end
    end
  end

            describe "BIP-39 Japanese vectors (NFKD code-path)" do
    for {v, idx} <- Enum.with_index(@japanese_vectors) do
      @v v
      @idx idx
      test "##{@idx} #{@v["description"]}" do
        words = String.split(@v["mnemonic"], "\u3000")
        seed = Bip39.mnemonic_to_seed(words, @v["passphrase"])
        assert Base.encode16(seed, case: :lower) == @v["seed_hex"]
      end
    end
  end

  describe "validate_mnemonic/1" do
    test "rejects a bad checksum" do
            words =
        "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon"
        |> String.split(" ")

      assert {:error, :bad_checksum} = Bip39.validate_mnemonic(words)
    end

    test "rejects an invalid word count" do
      assert {:error, :invalid_word_count} = Bip39.validate_mnemonic(~w(abandon abandon))
    end

    test "rejects a word not in the dictionary" do
      words =
        "notaword abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
        |> String.split(" ")

      assert {:error, :unknown_word} = Bip39.validate_mnemonic(words)
    end

    test "accepts the all-zero entropy mnemonic" do
      words =
        "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
        |> String.split(" ")

      assert :ok = Bip39.validate_mnemonic(words)
    end

    test "accepts the all-ones entropy mnemonic" do
      words =
        "zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo wrong"
        |> String.split(" ")

      assert :ok = Bip39.validate_mnemonic(words)
    end
  end

  describe "entropy_to_mnemonic/1" do
    test "rejects non-128-bit entropy" do
      assert_raise ArgumentError, fn -> Bip39.entropy_to_mnemonic(<<0::64>>) end
    end
  end

  describe "Hkdf.derive/4" do
    test "RFC 5869 test case 1 (SHA-256)" do
      ikm = Base.decode16!("0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B0B")
      salt = Base.decode16!("000102030405060708090A0B0C")
      info = Base.decode16!("F0F1F2F3F4F5F6F7F8F9")

      okm = Hkdf.derive(ikm, salt, info, 42)

      assert Base.encode16(okm, case: :lower) ==
               "3cb25f25faacd57a90434f64d0362f2a2d2d0a90cf1a5a4c5db02d56ecc4c5bf34007208d5b887185865"
    end
  end
end
